### PR TITLE
Remove package from inherited packages if project overrides it

### LINKED ISSUE #1151: "Inherited packages" tab shows non-inherited packages
https://build.opensuse.org/project/show/openSUSE:Factory:Staging:C shows cups as both package in the prj and also as inherited from ring1, which is wrong.

### PR DESCRIPTION
**Pull request by ChrisBr**
This will fix #1151

### CODE REVIEW

**Comment by ChrisBr** [comment]
I'll add some tests for it tomorrow!

**Comment by ChrisBr** [comment]
We should rewrite the whole function with a nice ActiveRecord query in the future!

@coolo @hennevogel please review!

**Comment by adrianschroeter** [comment]
can you add the webui commit tag also please

**Comment by coolo** [comment]
no idea what adrian is talking about, to me the implementation looks ok, but I lack a test case :)

**Comment by ChrisBr** [comment]
Added some tests and implemented it with an Active Record query.

**Comment by adrianschroeter** [comment]
the test case is not proving that the algorithm is doing the right thing (and I believe it doesn't). We look first deep. If this is wrong you would see no the same instance as the backend is using it.

So, you need a test case which has the following project links:
A -> B -> C
   -> D

If a package is in C and D, the instance from C needs to be taken for example.

**Comment by ChrisBr** [comment]
@adrianschroeter ahh you're right!

**Comment by ChrisBr** [comment]
@adrianschroeter I've adapted and added an additional test. I think that should cover now the scenario you mentioned?

I think the query is a better solution than using expand_all_packages which produces N+1 queries. The uniq preserves the ordering therefore it returns the packages from the linked_project with the highest position (if there are several).

File: src/api/test/unit/project_test.rb [hunk 1]
```diff
@@ -826,4 +826,76 @@ def test_cycle_handling
...
+                                linked_db_project_id: parent2.id,
+                                position: 2)
+
+    pack2 = parent2.packages.where(name: 'pack2').first
+    child.packages << pack2.dup
+
+    assert_equal child.packages_from_linked_projects,  parent1.packages + parent2.packages.where(name: 'pack2.linked')
+  end
+
+  test 'adrian test' do
```
Comment (bgeuken) [inline src/api/test/unit/project_test.rb]: Please add a proper test name here

File: src/api/app/models/project.rb [hunk 2]
```diff
@@ -1043,19 +1043,11 @@ def expand_maintained_projects
...
-    end
-  end
-
-  def linked_packages
-    expand_all_projects({}, false).reject{|project| project == self}.map {|project| project.packages.to_a }.flatten
+    all_packages = Package.
+        joins('LEFT OUTER JOIN linked_projects ON packages.project_id = linked_projects.linked_db_project_id').
+        where('linked_projects.db_project_id = ? AND name NOT IN (?)', id, packages.pluck(:name)).
+        order('linked_projects.position ASC, packages.name ASC')
+    all_packages.to_a.uniq { |p| p.name }
```
Comment (bgeuken) [inline src/api/app/models/project.rb]: If possible we should do the uniqness when we query the db.

File: src/api/app/controllers/webui/project_controller.rb [hunk 3]
```diff
@@ -901,7 +901,7 @@ def set_maintained_project
 
   def load_project_info
     find_maintenance_infos
-    @ipackages = @project.linked_packages.map { |package| [ package.name, package.project.name ] }
+    @ipackages = @project.packages_from_linked_projects.map { |package| [ package.name, package.project.name ] }
```
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: Given that packages_from_linked_projects is only used here, I would move the map to the other method and replace it by an SQL select.

File: src/api/test/unit/project_test.rb (hunk 1 above)
Comment (ChrisBr) [inline src/api/test/unit/project_test.rb]: haha, yes of course :) Thanks

File: src/api/app/models/project.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/project.rb]: Yeah, already tried it but unfortunately the database uniq doesn't preserve the ordering. We can talk later if you have a better idea for the query!

File: src/api/app/controllers/webui/project_controller.rb (hunk 3 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/project_controller.rb]: Hm, good idea!

File: src/api/app/models/project.rb (hunk 2 above)
Comment (bgeuken) [inline src/api/app/models/project.rb]: Hm, that's odd.

**Comment by ChrisBr** [comment]
@bgeuken fixed your suggestions! Tried a few things with the uniq but still no luck.

**Comment by bgeuken** [comment]
Looks good :+1:

**Pull request merged**